File: pkg/src/householder.c
/* $ID: householder.c, last updated 2026-01-12, F.Osorio */

#include "fastmatrix.h"

void
house_vec(double *x, int *n, double *u, double *tau)
{ /* Householder vector */
  int p = *n;

  *tau = FM_house(x, p, u);
}

void
house_mat(double *u, double *tau, int *n, double *mat)
{ /* Householder reflector */
  int p = *n;
  double beta = *tau;

  FM_householder_mat(u, beta, p, mat);
}

void 
house_prod_vec(double *x, int *n, double *u, double *tau)
{ /* applying a Householder reflection to a vector */
  int p = *n;
  double beta = *tau;

  FM_householder_vec(x, p, u, beta);
}

void
house_prod_mat(double *a, int *lda, int *nrow, int *ncol, int *job, double *u, double *tau)
{ /* applying a Householder reflection to a matrix */
  int n = *nrow, p = *ncol, task = *job;
  double beta = *tau;

  FM_householder_prod(a, *lda, n, p, task, u, beta);
}

double 
FM_house(double *x, int n, double *u)
{ /* creates the Householder vector */
  double beta, mu, v, s;

  if (n <= 1) /* quick return */
    return 0.0;

  s = FM_norm_sqr(x + 1, 1, n - 1);
  u[0] = 1.0;
  Memcpy(u + 1, x + 1, n - 1);

  v = x[0];
  if ((s == 0.0) && (v >= 0.0))
    return 0.0;
  else if ((s == 0.0) && (v < 0.0))
    return -2.0;
  
  mu = sqrt(SQR(v) + s);
  if (v <= 0.0)
    v = v - mu;
  else
    v = -s / (v + mu);
  u[0] = v;

  beta = 2.0 * SQR(v) / (s + SQR(v));
  BLAS1_scale(1.0 / v, u, 1, n);
  return beta;
}

void
FM_householder_mat(double *u, double tau, int n, double *mat)
{ /* creates the Householder matrix based on the vector u(1:n) */

  if (n <= 0) /* quick return */
    return;
  
  for (int i = 0; i < n; i++) {
    mat[i * (n + 1)] = 1.0 - tau * SQR(u[i]);
    for (int j = i + 1; j < n; j++) {
      *(mat + i + j * n) = -tau * u[i] * u[j];
      *(mat + j + i * n) = *(mat + i + j * n);
    }
  }
}

void
FM_householder_vec(double *x, int n, double *u, double tau)
{ /* applying a Householder reflection to a vector */
  double prod;

  if (n <= 0) /* quick return */
    return;

  prod = BLAS1_dot_product(u, 1, x, 1, n);
  tau *= -prod;
  BLAS1_axpy(tau, u, 1, x, 1, n);
}

void
FM_householder_prod(double *a, int lda, int n, int p, int job, double *u, double tau)
{ /* applying a Householder reflection to a matrix */
  char *trans = "T", *notrans = "N";
  double *v;

  if (n <= 0) /* quick return */
    return;

  if (job) { /* left: P %*% A */
    v = (double *) R_Calloc(p, double);
    BLAS2_gemv(1.0, a, lda, n, p, trans, u, 1, 0.0, v, 1);
  } else { /* right:  A %*% P */
    v = (double *) R_Calloc(n, double);
    BLAS2_gemv(1.0, a, lda, n, p, notrans, u, 1, 0.0, v, 1);
  }

  if (job) /* left: P %*% A */
    FM_rank1_update(a, lda, n, p, -tau, u, v);
  else /* right:  A %*% P */
    FM_rank1_update(a, lda, n, p, -tau, v, u);

  R_Free(v);
}

File: pkg/R/householder.R
## ID: householder.R, last updated 2026-01-13, F.Osorio

house <- function(x, matrix = FALSE)
{ ## Householder transformation
  if (!is.vector(x))
    stop("supply a vector-like 'x'")
  if (!is.numeric(x))
    stop("argument x is not a numeric vector")
  n <- length(x)

  z <- .C("house_vec",
          x = as.double(x),
          n = as.integer(n),
          u = double(n),
          tau = as.double(0))[c("u","tau")]
  if (matrix) {
    mat <- .C("house_mat",
              u = as.double(z$u),
              tau = as.double(z$tau),
              n = as.integer(n),
              mat = double(n * n))$mat
    mat <- matrix(mat, nrow = n, ncol = n)
  }

  o <- list(u = z$u, tau = z$tau)
  if (matrix)
    o <- mat
  o
}

house.prod <- function(a, x, side = "left")
{ ## apply a Householder reflection, 'house.prod' returns:
  ## y[,] <- H %*% a, side = "left", or
  ## y[,] <- a %*% H, side = "right",
  ## where H is the Householder transformation matrix.
  if (is.null(a))
    stop("argument 'x' is not supplied")
  if (is.data.frame(a))
    a <- as.matrix(a)
  if (!is.numeric(a))
    stop("'a' is not a numeric argument")
  if (is.vector(a)) {
    task <- 0
    n <- length(a)
  } else if (is.matrix(a)) {
    task <- 1
    da <- dim(a)
    n <- da[1]
    p <- da[2]
  } else 
    stop("the argument 'a' is not the right type.")

  if (!is.vector(x))
    stop("supply a vector-like 'x'")
  if (!is.numeric(x))
    stop("'x' is not a numeric argument")

  z <- house(x)
  if (task) {
    storage.mode(a) <- "double"
    switch(side,
           "left" = { # y[,] <- H %*% a
             if (n != length(x))
               stop("incompatible dimensions.")
             z <- .C("house_prod_mat",
                     a = a,
                     lda = as.integer(n),
                     nrow = as.integer(n),
                     ncol = as.integer(p),
                     job = as.integer(1),
                     u = as.double(z$u),
                     tau = as.double(z$tau))$a
             #z <- matrix(z, nrow = n, ncol = p)
           },
           "right" = { # y[,] <- a %*% H
             if (p != length(x))
               stop("incompatible dimensions.")
             z <- .C("house_prod_mat",
                     a = a,
                     lda = as.integer(n),
                     nrow = as.integer(n),
                     ncol = as.integer(p),
                     job = as.integer(0),
                     u = as.double(z$u),
                     tau = as.double(z$tau))$a 
           },
           stop("method not implemented."))
  } else {
    if (n != length(x))
      stop("incompatible dimensions.")
    z <- .C("house_prod_vec",
            x = as.double(x),
            n = as.integer(n),
            u = as.double(z$u),
            tau = as.double(z$tau))$x
  }
  z
}
